File: 3989-maximum-consistent-columns-in-a-grid/3989-maximum-consistent-columns-in-a-grid.cpp
class Solution {
public:
/*
so do a dp wherin dp[j] denotes the max possible columns if the last col is j

*/
vector<vector<int>>dp1;
vector<int>dp2;
bool valid(vector<vector<int>>& arr,int limit,  int a, int b){
    int rows=arr.size();
    bool ans=true;
    if(dp1[a][b]!=-1)return dp1[a][b];
    for(int i=0;i<rows;i++){
        if(abs(arr[i][b]-arr[i][a])>limit){
            ans=false;
            break;
        }
    }

    return dp1[a][b]=ans;
    
}
    
    int maxConsistentColumns(vector<vector<int>>& arr, int limit) {
        int rows=arr.size();
        int cols=arr[0].size();
        dp1.resize(cols,vector<int>(cols,-1));   
        dp2.resize(cols,1);
        
        
        for(int i=0;i<cols;i++){
            int prevmax=0;
            for(int j=cols-1;j>=0;j--){
                if(valid(arr,limit,j,i)){
                    prevmax=max(prevmax,dp2[j]);
                }
            }
            dp2[i]+=prevmax;
        }
        int ans=0;
        for(int i=0;i<cols;i++)ans=max(ans,dp2[i]);
        return ans-1;
    }
};

File: 3989-maximum-consistent-columns-in-a-grid/3989-maximum-consistent-columns-in-a-grid.java
import java.util.Arrays;

class Solution {
    // Helper method to check if column 'a' and column 'b' can be adjacent
    private boolean isCompatible(int[][] grid, int a, int b, int limit) {
        int rows = grid.length;
        for (int r = 0; r < rows; r++) {
            if (Math.abs(grid[r][b] - grid[r][a]) > limit) {
                return false;
            }
        }
        return true;
    }

    public int maxConsistentColumns(int[][] grid, int limit) {
        int m = grid.length;
        int n = grid[0].length;
        
        // dp[j] stores the max consistent columns ending at column j
        int[] dp = new int[n];
        Arrays.fill(dp, 1); // Every single column by itself is a valid sequence of length 1
        
        int maxColumns = 1;
        
        // Classic LIS approach
        for (int j = 0; j < n; j++) {
            for (int i = 0; i < j; i++) {
                if (isCompatible(grid, i, j, limit)) {
                    dp[j] = Math.max(dp[j], dp[i] + 1);
                }
            }
            maxColumns = Math.max(maxColumns, dp[j]);
        }
        
        return maxColumns;
    }
}